IPNI_FEDERATION.md

```diff
@@ -0,0 +1,425 @@
+2. Periodic snapshots - the process by which a node periodically captures an immutable snapshot of its providers along
+   with associated metadata.
+3. Snapshot exchange - the process by which each node propagates their latest snapshot to the rest of the network.
+4. Reconciliation - the process by which snapshots received from other nodes is reflected on to bring the network state
+   knowledge up-to-date.
+
+Finally, the specification expands on cases where conflict might occur and how a node should resolve it.
+
+### Initialization
+
+At the outset, every participating IPNI instance undergoes an initialization process. This involves setting up its
+vector clock to the default or baseline state and establishing secure communication channels with other participating
+instances. It's during this phase that instances also share their initial state and any configuration parameters crucial
+for the federation protocol. The initialization ensures a foundational layer of consistency, preparing the system for
+subsequent operations like snapshot exchange and reconciliation.
+
+### Periodic Snapshot
+
+Recognizing that continuous real-time synchronization could be resource-intensive and inefficient, the IPNI federation
+protocol employs a periodic snapshot mechanism. At predetermined intervals, or upon detecting significant changes, each
+IPNI instance captures a comprehensive snapshot of its index. This snapshot, which is essentially a compact
+representation of the instance's current state, serves as the primary data unit for synchronization among instances. It
+ensures that instances can exchange substantial amounts of data without overwhelming the network or causing excessive
+data traffic.
+
+The snapshot mechanism also offers a strategic advantage. In the event of any system failures or communication
+breakdowns, snapshots can be used to quickly restore or reconcile the state of an instance, minimizing data loss and
+ensuring rapid recovery.
+
+A snapshot consists of:
+
+* **Epoch**: a monotonically increasing value that represents the time at which snapshot was formed.
+* **Vector Clock**: the monotonically increasing vector clock value that corresponds to the IPNI instance.
+* **Provider to Ingest State Map**: a map of provider ID to the latest advertisement CID processed by the IPNI
+  instance, along with an optional advertisement height if known by the indexer. The height is provided as a _hint_ to
+  other indexers in order to aid the reconciliation and conflict resolution process.
+
+As the IPNI federation protocol continues to evolve, these specification components—initialization and periodic
+snapshots—act as the backbone, ensuring that the system remains resilient, efficient, and, most importantly,
+consistently synchronized across all instances. Future enhancements may introduce additional layers of optimization or
+fault-tolerance, but the essence will remain rooted in these foundational principles.
+
+### Snapshot Exchange
```

Comment:
As @hsanjuan mentioned elsewhere you might consider if a more prescriptive syncing scheme (e.g. having IPNI nodes post CRDT updates to a shared kv-store like ipfs-cluster uses https://github.com/ipfs/go-ds-crdt) would help you here.

This could reduce the amount of data emitted by snapshots as well as help with the synchronization among many nodes rather than leaving the "how to find the latest state" work up to the individual instances.